qmpi: failed to build in mpich-4.0

From discuss mailinglist:

> [quote]
Hi Mpich team,

I see the following issue with the latest 4.0 release of MPICH while adding QMPI support at configure time?

make[2]: Entering directory `/scratch1/05231/aruhela/libs/mpich/tar-mpich-4.0'
  CC       src/binding/c/attr/lib_libmpi_la-attr_delete.lo
src/binding/c/attr/attr_delete.c(63): error: attributes are not permitted in a function definition
  int QMPI_Attr_delete(QMPI_Context context, int tool_id, MPI_Comm comm, int keyval) MPICH_API_PUBLIC
                                                                                                                                                                 ^

compilation aborted for src/binding/c/attr/attr_delete.c (code 2)
make[2]: *** [src/binding/c/attr/lib_libmpi_la-attr_delete.lo] Error 1

Any suggestions to resolve the build error?

Amit Ruhela, Ph.D. 
HPC Software Tools Group 
Texas Advanced Computing Center 
The University of Texas at Austin

--- a/maint/local_python/binding_c.py
+++ b/maint/local_python/binding_c.py
@@ -159,7 +159,7 @@ def dump_mpir_impl_h(f):
         print("", file=Out)
         print("#endif /* MPIR_IMPL_H_INCLUDED */", file=Out)
 
-def get_qmpi_decl_from_func_decl(func_decl):
+def get_qmpi_decl_from_func_decl(func_decl, kind=""):
     if RE.match(r'(.*) (MPIX?_\w+)\((.*?)\)(.*)', func_decl):
         T, name, args, tail = RE.m.group(1,2,3,4)
     else:
@@ -170,11 +170,12 @@ def get_qmpi_decl_from_func_decl(func_decl):
     else:
         t = "%s Q%s(QMPI_Context context, int tool_id, %s)" % (T, name, args)
 
-    while RE.search(r'MPICH_ATTR_POINTER_WITH_TYPE_TAG\((\d+),(\d+)\)(.*)', tail):
-        i1, i2, tail = RE.m.group(1, 2, 3)
-        t += " MPICH_ATTR_POINTER_WITH_TYPE_TAG(%d,%d)" % (int(i1) + 2, int(i2) + 2)
+    if kind == 'proto':
+        while RE.search(r'MPICH_ATTR_POINTER_WITH_TYPE_TAG\((\d+),(\d+)\)(.*)', tail):
+            i1, i2, tail = RE.m.group(1, 2, 3)
+            t += " MPICH_ATTR_POINTER_WITH_TYPE_TAG(%d,%d)" % (int(i1) + 2, int(i2) + 2)
 
-    t += " MPICH_API_PUBLIC"
+        t += " MPICH_API_PUBLIC"
 
     return t
 
@@ -292,7 +293,7 @@ def dump_mpi_proto_h(f):
             m = re.match(r'[a-zA-Z0-9_]* ([a-zA-Z0-9_]*)\(.*', func_decl);
             if need_skip_qmpi(m.group(1)):
                 continue
-            func_decl = get_qmpi_decl_from_func_decl(func_decl)
+            func_decl = get_qmpi_decl_from_func_decl(func_decl, 'proto')
             dump_proto_line(func_decl, Out)
 
         print("", file=Out)

